### ISSUE TITLE
Some questions for inference and validation

### ISSUE THREAD

**Issue by albertofernandezvillan** [KevinMusgrave/pytorch-metric-learning#598]
I am training a model using an standard approach for both the trunk and the embedder:

```
class Net(nn.Module):
    def __init__(self, embedding_size):
        super(Net, self).__init__()
        self.trunk = torchvision.models.resnet18(pretrained=True)
        
        self.trunk_output_size = self.trunk.fc.in_features
        self.trunk.fc = torch.nn.Identity()
        self.embedder = torch.nn.Linear(self.trunk_output_size, embedding_size)

    def forward(self, x):
        x = self.trunk(x)
        x = self.embedder(x)
        return x
    
    def get_embedding(self, x):
        x = self.forward(x)
        # x = torch.nn.functional.normalize(x, p=2, dim=1)
        return x
```
I am using also ArcFaceLoss with a learning rate bigger for the loss_func (https://github.com/KevinMusgrave/pytorch-metric-learning/issues/534), which I think that is an standard approach:

```
loss_func = losses.ArcFaceLoss(num_classes=n_classes, embedding_size=embedding_size, scale=32)
learning_rate = 0.0001
optimizer = torch.optim.Adam(model.parameters(), lr=learning_rate)
W_optimizer = torch.optim.Adam(loss_func.parameters(), lr=learning_rate*10)

```
In my case, I have 7552 classes with 2 or 3 images per class, which I think it is similar to the conditions explained here (https://github.com/KevinMusgrave/pytorch-metric-learning/issues/351). This is why I have changed default `scale` value to a lower one. 

After training, I want to use the model for inference/prediction: given two images, if the similarity of the two images is high, I consider that they belong to the same class:

```
emb = model.get_embedding(img)
emb2 = model.get_embedding(img2)
cosine_sim = F.cosine_similarity(emb1, emb2)
```

`if consine_sim > threshold` they are for the same class. I have used cosine similarity as ArcFaceLoss uses this distance.

Moreover, for some “quick validation” while training, I calculate the embeddings for lets say 1000 test images, and for this images, I compute the closest embedding (using F.cosine_similarity()) shown above. If the closest embedding is for the same class, correct += 1. If not, incorrect += 1. At the end, I compute the “1 accuracy”. Also I compute the “5 accuracy” in the sense that if the class for every validation example is in the top 5 closest embeddings, I consider it correct. The code is more or less as follows:

```
def calculate_cos_sim(emb1, emb2):
    cosine_sim = F.cosine_similarity(emb1, emb2)
    return cosine_sim

def calculate_top_k_similar_embeddings(path_embedding_file, path_embedding_files, k):
    # priority queue to keep track of the k largest similarities encountered so far
    similarity_queue = []
    class_name, embedding = get_emb_class_from_file(path_embedding_file)

    for emb_file in path_embedding_files:
        curr_class_name, curr_emb = get_emb_class_from_file(emb_file)
        similarity = calculate_cos_sim(torch.Tensor(
            embedding).unsqueeze(0), torch.Tensor(curr_emb).unsqueeze(0))
        if len(similarity_queue) < k:
            heapq.heappush(similarity_queue, (similarity, curr_class_name))
        else:
            heapq.heappushpop(similarity_queue, (similarity, curr_class_name))

    for sim, pred_class in similarity_queue:
        if pred_class == class_name:
            return True
    return False

def calculate_accuracy_folder(folder, k):
    count_true, count_false = 0, 0
    embedding_files = os.listdir(folder)
    for embedding_file in tqdm(embedding_files):
        path_embedding_file = os.path.join(folder, embedding_file)
        path_embedding_files = get_files_except(folder, embedding_file)

        res = calculate_top_k_similar_embeddings(
            path_embedding_file, path_embedding_files, k)
        if res:
            count_true += 1
        else:
            count_false += 1

    accuracy = count_true / (count_true + count_false)
    return accuracy

def acc_val(model, data_transforms, folder):
    images_classes = get_paths("validation-files/images_classes_test_test.txt")
    write_embeddings(images_classes, data_transforms, model, folder)

    acc_1 = calculate_accuracy_folder(folder, k=1)
    acc_5 = calculate_accuracy_folder(folder, k=5)
    acc_10 = calculate_accuracy_folder(folder, k=10)

    return acc_1, acc_5, acc_10
```
`write_embeddings()` code calculate the embeddings and writes them to a file in disk (just for debugging at this moment)

So my questions are:

(1) Is the approach for inference/prediction OK?
(2) Also note the commented line in get_embedding() function. I don’t know if I should normalize it or not when using the embedding for inference/prediction.
(3) Is the approach for calculating the validation accuracy OK?
(4) For many "experiments" I find hard to choose valid parameters for the training (e.g. parameters in ArcFaceLoss and also in the optimizers). Is there a way with this library to see how the training is going? This is because for many experiments I see that the training loss is reducing (with fluctuations, but reducing) but the validation accuracy is decreasing.

**Comment by albertofernandezvillan** [comment]
Just an example of the metrics described above. Based on the chart, could be considered that the training is being OK? 
![image](https://user-images.githubusercontent.com/6097318/223651517-227a1061-d854-4246-bded-cf1bbb647d37.png)

**Comment by albertofernandezvillan** [comment]
> Aren't you already doing this by logging the training loss and validation accuracy?
Yes.

Closing this. Many thanks!
